### ISSUE TITLE
Amazon warning - Chrome version soon not supported any more

### ISSUE THREAD

**Issue by acroes** [tiliado/nuvolaruntime#497]
Nuvola is generating a warning when playing Amazon Music, about the Chrome version being too old, and soon not supported any longer: *“Chrome aktualisieren — Sie nutzen eine Version von Chrome, die bald nicht mehr unterstützt wird. Bitte aktualisieren Sie Ihren Browser.”*

![chrome-warning](https://user-images.githubusercontent.com/459562/51103134-f232b280-17e1-11e9-8fbe-92d1f128b300.png)

* I upgraded Nuvola to the latest version, also running up-to-date Ubuntu version.
* I use Gnome 3 (default Ubuntu) as desktop manager
* Ubuntu 18.10
* Amazon Cloud Player script: 5.8.0 (0-g00a02f7)
* Nuvola Apps Runtime: 4.15.3 (3-g5de0a02)
* Diorite: 4.15.21
* Web Engine: Chromium 71.0.3578.80

**Comment by a-m-s** [comment]
I see the same thing. Maybe Amazon isn't understanding the user agent correctly?

I do not see the message on music.amazon.co.uk in the Vivaldi browser (also based on chrome 71).

**Comment by acroes** [comment]
I do not get the warning message when I visit https://music.amazon.de/ directly from the chrome browser itself.
About Chromium: Version 71.0.3578.98 (Official Build) Built on Ubuntu , running on Ubuntu 18.10 (64-bit)

**Comment by fenryxo** [comment]
Hello @acroes. Thanks for reporting the issue. Amazon Music seems to have troubles with detection of the correct Chrome version in Nuvola.

Hello @a-m-s. The comparison of user agent string is here: 

* Chrome: `Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36`
* Vivaldi: `Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36 Vivaldi/2.2.1388.37`
* Nuvola: `Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.80 Nuvola/4.15.3 Safari/537.36`
* Nuvola with the patch below: `Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36 Nuvola/4.15.3`

```patch
diff --git a/metadata.in.json b/metadata.in.json
index b5d1fcc..6c73603 100644
--- a/metadata.in.json
+++ b/metadata.in.json
@@ -13,6 +13,8 @@
   "allowed_uri": "^https?://(music.amazon.[a-z.]+/#?.*|www.amazon.[a-z.]+/[ga]p/.*)$",
   "#requirements": "https://github.com/tiliado/nuvola-app-amazon-cloud-player/issues/33",
   "requirements": "Chromium[65] Codec[MP3] Feature[MSE] Feature[Flash]",
+  "#user_agent": "https://github.com/tiliado/nuvolaruntime/issues/497",
+  "user_agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36",
   "build": {
     "icons": [
       "src/icon.svg SCALABLE 64 128 256",
```

Neither of the Nuvola user agents is good for Amazon. I need to modify Nuvola [not to add `Nuvola/4.x.x` to the script-specified user agent](https://github.com/tiliado/nuvolaruntime/blob/master/src/nuvolakit-runner/cef/CefOptions.vala#L105) to see whether a user agent string without `Nuvola` is ok.

The check is probably in [`WebPlayer.main.4af4b3f3b44dff9fa607._CB460198379_.js`](https://images-eu.ssl-images-amazon.com/images/G/01/digital/music/player/web/assets/WebPlayer.main.4af4b3f3b44dff9fa607._CB460198379_.js) ([a pretty-printed version](https://github.com/tiliado/nuvolaruntime/files/2754977/amazon_web_player.js.txt)).

**Comment by fenryxo** [comment]
Neither `Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36` does work.

**Comment by fenryxo** [comment]
There is a test whether EME (Encrypted Media Extension) is supported. Amazon Music obviously plans to drop the Flash-based player :tada: and rely on HTML5 Audio with Widevine DRM.

@a-m-s, could you test the patch below? Remember to regenerate `metadata.json` with `./configure`. If it works, mention tiliado/nuvola-app-amazon-cloud-player#33 in the commit message too.

```patch
diff --git a/metadata.in.json b/metadata.in.json
index b5d1fcc..cc21663 100644
--- a/metadata.in.json
+++ b/metadata.in.json
@@ -11,8 +11,7 @@
   "dark_theme": true,
   "license": "2-Clause BSD-license",
   "allowed_uri": "^https?://(music.amazon.[a-z.]+/#?.*|www.amazon.[a-z.]+/[ga]p/.*)$",
-  "#requirements": "https://github.com/tiliado/nuvola-app-amazon-cloud-player/issues/33",
-  "requirements": "Chromium[65] Codec[MP3] Feature[MSE] Feature[Flash]",
+  "requirements": "Chromium[71] Codec[MP3] Feature[MSE] Feature[Widevine]",
   "build": {
     "icons": [
       "src/icon.svg SCALABLE 64 128 256",
```

**Comment by a-m-s** [comment]
The patch seems to work, although I couldn't get Nuvola to run at all at first. Apparently something wasn't as updated as I thought it was.

Anyway, I guess Amazon finally got around to replacing Flash. Should I commit right away, or is there some reason to wait?

**Comment by a-m-s** [comment]
I don't have any messages and I can't reproduce the issue now.

The switch to Widevine is now pushed. This issue looks fixed to me.

**Comment by acroes** [comment]
I am away from my PC for the next few days, so can unfortunately not test
the fix myself. Will try after my return.
Thanks so far anyway. Great product, great support.

Regards,
Arnaud

**Comment by fenryxo** [comment]
Flatpak packages have been rebuilt and pushed. It's Amazon script version 5.8.2 (2-gb9ce8a9). @acroes, could you verify that the issue is fixed when you have time? Thanks.

**Comment by acroes** [comment]
I have upgrade to 5.8.2 (2-gb9ce8a9), which triggered the installation
of widevine, and can now confirm the chrome error no longer appears.
Great job. Thank you !!

**Comment by fenryxo** [comment]
Great :-)

**Comment by fenryxo** [comment]
Released with **Nuvola 4.16.0**, more info: [Nuvola 4.16 Brings Libre.FM Scrobbling, Ivoox Service, and more](https://medium.com/nuvola-news/nuvola-4-16-86a089068784)
